File: examples/primitives/transaction/eip7702-authorization.ts
/**
 * EIP-7702 Authorization Transaction Example
 *
 * Demonstrates EOA delegation allowing externally-owned accounts
 * to temporarily execute as smart contracts:
 * - Creating authorization signatures
 * - Batched operations
 * - Social recovery patterns
 * - Gas abstraction
 */

import * as Address from "../../../src/primitives/Address/index.js";
import * as Hex from "../../../src/primitives/Hex/index.js";
import * as Transaction from "../../../src/primitives/Transaction/index.js";

const authorization: Transaction.Authorization = {
	chainId: 1n,
	address: Address.from("0x1111111111111111111111111111111111111111"), // Smart wallet contract
	nonce: 0n, // EOA nonce when signing authorization
	yParity: 0,
	r: Hex.toBytes(
		"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
	),
	s: Hex.toBytes(
		"0xfedcba0987654321fedcba0987654321fedcba0987654321fedcba0987654321",
	),
};

const eip7702Tx: Transaction.EIP7702 = {
	type: Transaction.Type.EIP7702,
	chainId: 1n,
	nonce: 0n,
	maxPriorityFeePerGas: 1_000_000_000n,
	maxFeePerGas: 20_000_000_000n,
	gasLimit: 100_000n,
	to: Address.from("0x742d35Cc6634C0532925a3b844Bc9e7595f51e3e"),
	value: 0n,
	data: Hex.toBytes(`0xa9059cbb${"00".repeat(64)}`), // Contract call data
	accessList: [],
	authorizationList: [authorization],
	yParity: 0,
	r: Hex.toBytes(`0x${"00".repeat(32)}`),
	s: Hex.toBytes(`0x${"00".repeat(32)}`),
};

const batchWalletAddress = Address.from(
	"0x2222222222222222222222222222222222222222",
);

const batchAuth: Transaction.Authorization = {
	chainId: 1n,
	address: batchWalletAddress, // Contract with executeBatch() function
	nonce: 0n,
	yParity: 0,
	r: Hex.toBytes(`0x${"00".repeat(32)}`),
	s: Hex.toBytes(`0x${"00".repeat(32)}`),
};

const multiAuthTx: Transaction.EIP7702 = {
	type: Transaction.Type.EIP7702,
	chainId: 1n,
	nonce: 0n,
	maxPriorityFeePerGas: 1_000_000_000n,
	maxFeePerGas: 20_000_000_000n,
	gasLimit: 200_000n,
	to: null,
	value: 0n,
	data: new Uint8Array(),
	accessList: [],
	authorizationList: [
		{
			chainId: 1n,
			address: Address.from("0x1111111111111111111111111111111111111111"),
			nonce: 5n,
			yParity: 0,
			r: Hex.toBytes(`0x${"00".repeat(32)}`),
			s: Hex.toBytes(`0x${"00".repeat(32)}`),
		},
		{
			chainId: 1n,
			address: Address.from("0x2222222222222222222222222222222222222222"),
			nonce: 10n,
			yParity: 1,
			r: Hex.toBytes(`0x${"00".repeat(32)}`),
			s: Hex.toBytes(`0x${"00".repeat(32)}`),
		},
	],
	yParity: 0,
	r: Hex.toBytes(`0x${"00".repeat(32)}`),
	s: Hex.toBytes(`0x${"00".repeat(32)}`),
};
for (let i = 0; i < multiAuthTx.authorizationList.length; i++) {
	const auth = multiAuthTx.authorizationList[i];
}

const recoveryContractAddr = Address.from(
	"0x3333333333333333333333333333333333333333",
);

const gaslessContractAddr = Address.from(
	"0x4444444444444444444444444444444444444444",
);

const spendingLimitAddr = Address.from(
	"0x5555555555555555555555555555555555555555",
);

File: examples/primitives/transaction/eip7702-authorization.zig
// EIP-7702 Authorization Transaction Example
//
// Demonstrates EOA delegation - allowing externally-owned accounts
// to temporarily execute as smart contracts

const std = @import("std");
const primitives = @import("primitives");

const Address = primitives.Address.Address;

pub fn main() !void {
    std.debug.print("\n=== EIP-7702 Authorization Transaction Examples ===\n\n", .{});

    // Example 1: Basic authorization structure
    std.debug.print("1. Basic EIP-7702 Authorization\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    const smart_wallet = try Address.fromHex("0x1111111111111111111111111111111111111111");

    std.debug.print("Authorization Structure:\n", .{});
    std.debug.print("  chainId: 1\n", .{});
    std.debug.print("  address: 0x{X} (smart wallet)\n", .{smart_wallet.bytes});
    std.debug.print("  nonce: 0 (EOA nonce at signing time)\n", .{});
    std.debug.print("  yParity: 0 or 1\n", .{});
    std.debug.print("  r: 32-byte signature\n", .{});
    std.debug.print("  s: 32-byte signature\n", .{});
    std.debug.print("\n", .{});

    std.debug.print("Transaction Lifecycle:\n", .{});
    std.debug.print("  Before TX: EOA.code = empty\n", .{});
    std.debug.print("  During TX: EOA.code = DELEGATECALL(0x{X})\n", .{smart_wallet.bytes});
    std.debug.print("  After TX: EOA.code = empty (reverted)\n", .{});
    std.debug.print("\n", .{});

    // Example 2: Use cases
    std.debug.print("2. Common Use Cases\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    std.debug.print("Batched Transactions:\n", .{});
    std.debug.print("  Delegate to batch wallet contract\n", .{});
    std.debug.print("  Execute multiple calls in one TX\n", .{});
    std.debug.print("  Example: Transfer token A + token B + swap\n", .{});
    std.debug.print("\n", .{});

    std.debug.print("Social Recovery:\n", .{});
    std.debug.print("  Delegate to recovery contract\n", .{});
    std.debug.print("  Guardians can initiate recovery\n", .{});
    std.debug.print("  Restore access to lost accounts\n", .{});
    std.debug.print("\n", .{});

    std.debug.print("Gas Abstraction:\n", .{});
    std.debug.print("  Delegate to gasless wallet contract\n", .{});
    std.debug.print("  User pays in tokens\n", .{});
    std.debug.print("  Relayer pays gas in ETH\n", .{});
    std.debug.print("\n", .{});

    std.debug.print("Spending Limits:\n", .{});
    std.debug.print("  Delegate to limit enforcement contract\n", .{});
    std.debug.print("  Enforce daily/per-tx limits\n", .{});
    std.debug.print("  Enhanced security for EOAs\n", .{});
    std.debug.print("\n", .{});

    // Example 3: Security considerations
    std.debug.print("3. Security Considerations\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    std.debug.print("Nonce Protection:\n", .{});
    std.debug.print("  Authorization includes EOA nonce\n", .{});
    std.debug.print("  If nonce mismatch → rejected\n", .{});
    std.debug.print("  Prevents replay attacks\n", .{});
    std.debug.print("\n", .{});

    std.debug.print("Chain ID Binding:\n", .{});
    std.debug.print("  Authorization includes chain ID\n", .{});
    std.debug.print("  Cannot replay on different chains\n", .{});
    std.debug.print("  Cross-chain protection\n", .{});
    std.debug.print("\n", .{});

    std.debug.print("Temporary Delegation:\n", .{});
    std.debug.print("  Delegation ONLY during transaction\n", .{});
    std.debug.print("  EOA automatically reverts after TX\n", .{});
    std.debug.print("  No permanent state change\n", .{});
    std.debug.print("\n", .{});

    std.debug.print("Contract Trust:\n", .{});
    std.debug.print("  ⚠️  Delegated contract has full control\n", .{});
    std.debug.print("  ⚠️  Can drain funds during TX\n", .{});
    std.debug.print("  ✓  Only delegate to audited contracts\n", .{});
    std.debug.print("\n", .{});

    // Example 4: Comparison with Account Abstraction
    std.debug.print("4. EIP-7702 vs Account Abstraction\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    std.debug.print("EIP-7702:\n", .{});
    std.debug.print("  Account Type: EOA (temporary)\n", .{});
    std.debug.print("  Deployment: None required\n", .{});
    std.debug.print("  Duration: Per transaction\n", .{});
    std.debug.print("  Compatibility: Full EOA compat\n", .{});
    std.debug.print("  Cost: Lower\n", .{});
    std.debug.print("\n", .{});

    std.debug.print("ERC-4337:\n", .{});
    std.debug.print("  Account Type: Smart contract\n", .{});
    std.debug.print("  Deployment: Required\n", .{});
    std.debug.print("  Duration: Permanent\n", .{});
    std.debug.print("  Compatibility: New type\n", .{});
    std.debug.print("  Cost: Higher\n", .{});
    std.debug.print("\n", .{});

    // Example 5: Authorization signing
    std.debug.print("5. Authorization Signing Process\n", .{});
    std.debug.print("-" ** 50 ++ "\n", .{});

    std.debug.print("Signing Hash Structure:\n", .{});
    std.debug.print("  MAGIC = 0x05 (EIP-7702 authority type)\n", .{});
    std.debug.print("  authSigningHash = keccak256(\n", .{});
    std.debug.print("    MAGIC || rlp([chainId, address, nonce])\n", .{});
    std.debug.print("  )\n", .{});
    std.debug.print("\n", .{});

    std.debug.print("Sign with EOA private key:\n", .{});
    std.debug.print("  (r, s, yParity) = sign(authSigningHash, eoaPrivKey)\n", .{});
    std.debug.print("\n", .{});

    std.debug.print("Authorization is valid when:\n", .{});
    std.debug.print("  ✓ Signature is valid\n", .{});
    std.debug.print("  ✓ chainId matches current chain\n", .{});
    std.debug.print("  ✓ nonce matches EOA nonce\n", .{});
    std.debug.print("  ✓ Recovered address matches EOA\n", .{});

    std.debug.print("\n=== Example Complete ===\n\n", .{});
}
